cfs_config_util/cfs.py

```diff
@@ -376,24 +387,42 @@ def req_payload(self):
         """dict: a dict containing just the layers key used to update in requests"""
         return {'layers': [layer.req_payload for layer in self.layers]}
 
-    def save_to_cfs(self, name=None):
+    def save_to_cfs(self, name=None, overwrite=True):
         """Save the configuration to CFS, optionally with a new name.
 
         Args:
             name (str or None): the name to save as. Required if this
                 configuration does not yet have a name.
+            overwrite (bool): if True, silently overwrite an existing CFS
+                configuration given by `name`. If False, raise a
+                CFSConfigurationError if a configuration with `name` already
+                exists
 
         Returns:
             CFSConfiguration: the new configuration that was saved to CFS
 
         Raises:
             CFSConfigurationError: if there is a failure saving the configuration
-                to CFS.
+                to CFS, or if a CFS configuration exists and `overwrite` is False
         """
         if not name and not self.name:
             raise ValueError('A name must be specified for the CFS configuration.')
         name = name or self.name
 
+        # If overwriting is disabled, throw an error if we try to overwrite a
+        # configuration
+        if not overwrite:
+            try:
+                response = self._cfs_client.get('v2', 'configurations', name, raise_not_ok=False)
+                if response.status_code != 404:
+                    raise CFSConfigurationError(f'A configuration named {name} already exists '
+                                                f'and may not be overwritten.')
```

Comment:
This condition will raise the `CFSConfigurationError` with a message that the configuration already exists when the `response.status_code` is not ok, but not 404. I think we really only want to raise an "Already exists" error when the response is ok and some other error when the response is not ok but not 404. Something like:

```suggestion
                # If the response is okay, it already exists
                if response.ok:
                    raise CFSConfigurationError(f'A configuration named {name} already exists '
                                                f'and may not be overwritten.')
                # Response of 404 Not Found is okay
                elif response.status_code != 404:
                    raise APIError(f'Failed to retrieve CFS configuration "{self.name}": '
                                   f'CFS returned status code {response.status_code}: {response.reason}')
```

Reply:
As an aside, there are a few places we should do something like this in the `sat` package as well. Specifically, I have some TODO comments about this in `sat bootprep`.

Reply:
I agree, I believe there's an instance of this in `sat swap blade` too relating to waiting for NodeBMCs to be discovered. I can create a JIRA to port the `APIClient` API changes back to SAT and use that for those two instances.

Reply:
Thanks!